src/shader/ShaderDynamic.js

```diff
@@ -1350,11 +1347,24 @@ x3dom.shader.DynamicShader.prototype.generateFragmentShader = function ( gl, pro
 
     shader += "color = " + x3dom.shader.encodeGamma( properties, "color" ) + ";\n";
 
-    //Fog
-    if ( properties.FOG && !properties.SHADOW )
+    if ( properties.FOG )
     {
-        shader += "float f0 = calcFog(fragEyePosition);\n";
-        shader += "color.rgb = fogColor * (1.0-f0) + f0 * (color.rgb);\n";
+        shader +=   "  float f0 = calcFog(fragEyePosition);\n" +
+                    //"  color.rgb = fogColor * (1.0-f0) + f0 * (color.rgb);\n";
+                    // original fog above, or optional Noise and Height below
+                    "  if (fogNoise > 0.0) f0 += mix( 0.0, 0.1, noise3d(fragPositionWS.xyz * fogNoise) );\n" +
+                    "    vec3 fcol = fogColor * (1.0-f0) + f0 * (color.rgb);\n" +
+                    "  if (fogHeight == 0.0) {\n" +
+                    "    color.rgb = mix( fcol, color.rgb, f0 );\n" +
+                    "  } else if (fogHeight > 0.0) {\n" +
+                    "    float f1 = clamp( 1.0 - ((abs(fogHeight)-fragPositionWS.y) / abs(fogHeight)), 0.0, 1.0);\n" +
+                    "    color.rgb = mix( fcol, color.rgb, max(f1,f0) );\n" +
+                    "  } else if (fogHeight < 0.0) {\n" +
+                    "    float f1 = clamp(       ((abs(fogHeight)-fragPositionWS.y) / abs(fogHeight)), 0.0, 1.0);\n" +
```

Comment:
Since fogHeight negative, this is:
`(-fogHeight-fragPositionWS.y)/-fogHeight =
(fogHeight+fragPositionWS,y)/fogHeight`
simplified. 
May need more thinking.